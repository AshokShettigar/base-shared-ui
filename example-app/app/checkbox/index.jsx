import { Button, CheckBox, Field, Input } from 'platform-common-ui';
import BaseComponentView from 'app/components/base-component-view';

function cloneOptions(origOptions, replaceOption, replaceIndex) {
    const options = [];
    for (let i = 0; i < origOptions.length; i++) {
        options.push(replaceIndex === i ? replaceOption : origOptions[i]);
    }
    return options;
}

export default class InputExample extends BaseComponentView {
    constructor(props) {
        super(props);
        this.state = {
            hasError: false,
            isDisabled: false,
            isRequired: false,
            newOption: {
                isChecked: false,
                isDisabled: false,
                hideLabel: false,
                label: '',
                value: ''
            },
            options: [{
                hideLabel: false,
                isChecked: false,
                isDisabled: false,
                label: 'Option 1',
                value: 'opt1'
            }, {
                hideLabel: false,
                isChecked: false,
                isDisabled: false,
                label: 'Option 2',
                value: 'opt2'
            }]
        };
        this.onAddNewOption = () => {
            const options = this.state.options.concat(this.state.newOption);
            this.setState({
                options,
                newOption: {
                    label: '',
                    value: '',
                    isChecked: false,
                    isDisabled: false
                }
            });
        };
        this.onDisabledChange = () => {
            this.setState({ isDisabled: !this.state.isDisabled });
        };
        this.onHasErrorChange = () => {
            this.setState({ hasError: !this.state.hasError });
        };
        this.onNewOptionHideLabelChange = () => {
            const newOption = {
                ...this.state.newOption,
                hideLabel: !this.state.newOption.hideLabel
            };
            this.setState({ newOption });
        };
        this.onNewOptionIsCheckedChange = () => {
            const newOption = {
                ...this.state.newOption,
                isChecked: !this.state.newOption.isChecked
            };
            this.setState({ newOption });
        };
        this.onNewOptionIsDisabledChange = () => {
            const newOption = {
                ...this.state.newOption,
                isDisabled: !this.state.newOption.isDisabled
            };
            this.setState({ newOption });
        };
        this.onNewOptionLabelChange = label => {
            const newOption = {
                ...this.state.newOption,
                label
            };
            this.setState({ newOption });
        };
        this.onNewOptionValueChange = value => {
            const newOption = {
                ...this.state.newOption,
                value
            };
            this.setState({ newOption });
        };
        this.onOptionIsCheckedChange = index => () => {
            const option = this.state.options[index];
            const newOption = {
                ...option,
                isChecked: !option.isChecked
            };

            const options = cloneOptions(this.state.options, newOption, index);
            this.setState({ options });
        };
        this.onOptionIsDisabledChange = index => () => {
            const option = this.state.options[index];
            const newOption = {
                ...option,
                isDisabled: !option.isDisabled
            };

            const options = cloneOptions(this.state.options, newOption, index);
            this.setState({ options });
        };
        this.onOptionHideLabelChange = index => () => {
            const option = this.state.options[index];
            const newOption = {
                ...option,
                hideLabel: !option.hideLabel
            };

            const options = cloneOptions(this.state.options, newOption, index);
            this.setState({ options });
        };
        this.onOptionLabelChange = index => label => {
            const option = this.state.options[index];
            const newOption = {
                ...option,
                label
            };

            const options = cloneOptions(this.state.options, newOption, index);
            this.setState({ options });
        };
        this.onOptionValueChange = index => value => {
            const option = this.state.options[index];
            const newOption = {
                ...option,
                value
            };

            const options = cloneOptions(this.state.options, newOption, index);
            this.setState({ options });
        };
        this.onRemoveOption = index => () => {
            const options = [].concat(this.state.options);
            options.splice(index, 1);
            this.setState({ options });
        };
        this.onRequiredChange = () => {
            this.setState({ isRequired: !this.state.isRequired });
        };
    }

    getOptions() {
        return (
            <div>
                <Field label="Field Options">
                    <CheckBox
                        id="fieldDisable"
                        isChecked={this.state.isDisabled}
                        onChange={this.onDisabledChange}
                        value="disable"
                    >
                        Disable field
                    </CheckBox>
                    <CheckBox
                        id="fieldRequired"
                        isChecked={this.state.isRequired}
                        onChange={this.onRequiredChange}
                        value="required"
                    >
                        Enable required
                    </CheckBox>
                    <CheckBox
                        id="fieldHasError"
                        isChecked={this.state.hasError}
                        onChange={this.onHasErrorChange}
                        value="error"
                    >
                        Set error
                    </CheckBox>
                </Field>
                <Field label="Options List">
                    {
                        this.state.options.map((option, index) => this.renderOption(
                            option.label,
                            option.value,
                            option.isChecked,
                            option.isDisabled,
                            option.hideLabel,
                            index))
                    }
                    <Field className="app-option" label="New option">
                        <Input
                            label="Label"
                            id={this.state.newOption.label}
                            onChange={this.onNewOptionLabelChange}
                            value={this.state.newOption.label}
                        />
                        <Input
                            label="Value"
                            id={this.state.newOption.value}
                            onChange={this.onNewOptionValueChange}
                            value={this.state.newOption.value}
                        />
                        <CheckBox
                            id="addNewCheck"
                            isChecked={this.state.newOption.isChecked}
                            onChange={this.onNewOptionIsCheckedChange}
                            value={this.state.newOption.value}
                        >
                            Check
                        </CheckBox>
                        <CheckBox
                            id="addNewDisable"
                            isChecked={this.state.newOption.isDisabled}
                            onChange={this.onNewOptionIsDisabledChange}
                            value={this.state.newOption.value}
                        >
                            Disable
                        </CheckBox>
                        <CheckBox
                            id="addNewHideLabel"
                            isChecked={this.state.newOption.hideLabel}
                            onChange={this.onNewOptionHideLabelChange}
                            value={this.state.newOption.value}
                        >
                            Hide Label
                        </CheckBox>
                        <Button onClick={this.onAddNewOption}>Add new</Button>
                    </Field>
                </Field>
            </div>
        );
    }

    renderComponent() {
        return (
            <Field
                hasError={this.state.hasError}
                isDisabled={this.state.isDisabled}
                isRequired={this.state.isRequired}
                label="Field option example"
            >
                {
                    this.state.options.map((option, index) => (
                        <CheckBox
                            hideLabel={option.hideLabel}
                            id={option.value}
                            isChecked={option.isChecked}
                            isDisabled={option.isDisabled}
                            key={option.value}
                            onChange={this.onOptionIsCheckedChange(index)}
                            value={option.value}
                        >
                            {option.label}
                        </CheckBox>
                    ))
                }
            </Field>
        );
    }

    renderOption(label, value, isChecked, isDisabled, hideLabel, index) {
        return (
            <Field className="app-option" label={label} key={value}>
                <Input
                    label="Label"
                    id={label}
                    onChange={this.onOptionLabelChange(index)}
                    value={label}
                />
                <Input
                    label="Value"
                    id={value}
                    onChange={this.onOptionValueChange(index)}
                    value={value}
                />
                <CheckBox
                    id={`ck_${value}`}
                    isChecked={isChecked}
                    onChange={this.onOptionIsCheckedChange(index)}
                    value={value}
                >
                    Check
                </CheckBox>
                <CheckBox
                    id={`di_${value}`}
                    isChecked={isDisabled}
                    onChange={this.onOptionIsDisabledChange(index)}
                    value={value}
                >
                    Disable
                </CheckBox>
                <CheckBox
                    id={`hl_${value}`}
                    isChecked={hideLabel}
                    onChange={this.onOptionHideLabelChange(index)}
                    value={value}
                >
                    Hide Label
                </CheckBox>
                <Button onClick={this.onRemoveOption(index)}>Remove</Button>
            </Field>
        );
    }
}
