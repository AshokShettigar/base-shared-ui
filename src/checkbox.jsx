import classNames from 'classnames';
import PropTypes from 'prop-types';
import { onChangeInput } from 'platform-common-ui/input';

// eslint-disable-next-line import/prefer-default-export
export function CheckBox(props) {
    const className = classNames({
        'cn-field__opt': true,
        'cn-field__opt--is-checked': props.isChecked
    }, props.className);

    const textClassName = classNames({
        'cn-field__text': true,
        a11y: props.hideLabel
    });

    return (
        <label className={className} htmlFor={props.id}>
            <input
                checked={props.isChecked}
                className="cn-field__checkradio"
                disabled={props.isDisabled}
                id={props.id}
                onChange={onChangeInput(props.onChange)}
                type={props.type}
                value={props.value}
            />
            <span className={textClassName}>{props.children}</span>
        </label>
    );
}

CheckBox.propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    hideLabel: PropTypes.bool.isRequired,
    id: PropTypes.string.isRequired,
    isChecked: PropTypes.bool.isRequired,
    isDisabled: PropTypes.bool.isRequired,
    onChange: PropTypes.func.isRequired,
    type: PropTypes.oneOf(['checkbox', 'radio']).isRequired,
    value: PropTypes.string.isRequired
};

CheckBox.defaultProps = {
    hideLabel: false,
    isChecked: false,
    isDisabled: false,
    type: 'checkbox'
};
