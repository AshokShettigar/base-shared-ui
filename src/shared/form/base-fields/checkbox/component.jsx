import { PureComponent } from 'react';
import { string, bool, func, node } from 'prop-types';
import classNames from 'classnames';

import { FieldWithError } from '../index';
import { withRender } from '../../../shared';
import styles from './component.scss';
import { Icon } from '../../../icon';

class Checkbox extends PureComponent {
    static propTypes = {
        id: string,
        name: string,
        value: string,
        children: node,
        error: string,

        autoFocus: bool,
        readOnly: bool,
        checked: bool,
        multiple: bool,
        disabled: bool,
        indeterminate: bool,

        className: string,
        labelClassName: string,

        innerRef: func,
        onChange: func
    };

    static defaultProps = {
        id: undefined,
        name: undefined,
        value: undefined,
        children: undefined,
        error: undefined,

        autoFocus: false,
        readOnly: false,
        checked: false,
        multiple: false,
        disabled: false,
        indeterminate: false,

        className: undefined,
        labelClassName: undefined,

        innerRef: undefined,
        onChange: undefined
    };

    onChange = () => {
        if (this.props.onChange) {
            this.props.onChange({
                name: this.props.name,
                value: this.props.value,
                checked: this.checkboxRef.checked
            });
        }
    };

    checkboxRef = null;

    setRef = el => {
        this.checkboxRef = el;

        if (this.props.innerRef) {
            this.props.innerRef(el);
        }
    };

    render() {
        const className = classNames(styles.root, this.props.className, {
            [styles.multiple]: this.props.multiple,
            [styles.disabled]: this.props.disabled,
            [styles.readOnly]: this.props.readOnly,
            [styles.indeterminate]: this.props.indeterminate,
            [styles.checked]: this.props.checked
        });

        return (
            <span className={className}>
                <FieldWithError error={this.props.error}>
                    <label htmlFor={this.props.id} className={styles.label}>
                        <input
                            hidden
                            type="checkbox"
                            autoFocus={this.props.autoFocus} // eslint-disable-line
                            name={this.props.name}
                            value={this.props.value}
                            checked={this.props.checked}
                            disabled={this.props.disabled}
                            ref={this.setRef}
                            onChange={this.onChange}
                        />
                        <span className={styles.checkbox}>
                            {this.props.indeterminate && <i className={styles.indeterminateGlyph} />}
                            {this.props.checked && !this.props.indeterminate &&
                                <span className={styles.checkmarkGlyph}>
                                    <Icon glyph="checkmark" size="mini" className={styles.icon} />
                                </span>
                            }
                        </span>
                        {this.props.children &&
                            <span
                                tabIndex={-1}
                                role="radio"
                                aria-checked={this.props.checked}
                                className={classNames(styles.labelText, this.props.labelClassName)}
                            >
                                {this.props.children}
                            </span>
                    }
                    </label>
                </FieldWithError>
            </span>
        );
    }
}

export default withRender()(Checkbox);
