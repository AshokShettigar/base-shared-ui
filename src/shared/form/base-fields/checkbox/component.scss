@import '../../../scss/variables';

.root {
  display: inline-flex;
  font-family: $ff-open-sans;
  font-size: $fs-tiny;
  color: $c-raisin-black;
}

.label {
  display: flex;
  cursor: pointer;
  margin: 0;
}

.labelText {
  line-height: 12px;
  margin-left: 5px;
  color: $c-raisin-black;
  user-select: none;
  display: flex;
}

.checkbox {
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  min-width: 12px;
  height: 12px;
  border: 1px solid $c-gainsboro;
  border-radius: 2px;
  background-color: $c-white;
  z-index: 1;
}

.checkmarkGlyph,
.indeterminateGlyph {
  width: 10px;
  height: 10px;
  position: relative;

  &::after {
    position: absolute;
    top: 50%;
    left: 50%;
  }
}

.checkmarkGlyph {
  &::after {
    margin-top: -5px;
    margin-left: -5px;
  }
}

.indeterminateGlyph {
  &::after {
    content: ' ';
    width: 8px;
    height: 1px;
    background: currentColor;
    margin-left: -4px;
  }
}

.multiple {
  position: relative;

  &::after {
    content: ' ';
    position: absolute;
    width: 12px;
    height: 12px;
    bottom: -2px;
    left: -2px;
    border: 1px solid $c-gainsboro;
    border-top-color: transparent;
    border-right-color: transparent;
    display: block;
    border-radius: 2px;
  }

  &.checked::after {
    border-left: 1px solid $c-french-blue;
    border-bottom: 1px solid $c-french-blue;
  }
}

.checked, .indeterminate {
  .checkbox {
    background-color: $c-french-blue;
    border: 1px solid transparent;
    color: $c-white;
  }
}

.disabled, .readOnly {
  .label {
    cursor: default;
  }

  .checkbox {
    border: 1px solid $c-gainsboro;
    background-color: $c-white-smoke;
    color: $c-gainsboro;
  }
}

.icon {
  display: flex;
}
