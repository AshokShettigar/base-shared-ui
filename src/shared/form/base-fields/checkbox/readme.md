## Checkbox
#### Usage
```JSX
import { shared } from 'platform-common-ui';
const { Checkbox } = shared;
```
```JSX
<Checkbox />
```

#### Description
Checkbox component permits the user to make a binary choice.

#### Properties
Name                  | Type               | Required | Default     | Description       
:---------------------| :----------:       | :------: | :-----:     | :-----------
id                    | `string`           | `-`      | undefined   | Setup the Checkbox's id
name                  | `string`           | `-`      | undefined   | Setup the Checkbox's name
value                 | `string`           | `-`      | undefined   | Setup the Checkbox's value
children              | `node`             | `-`      | undefined   | Displays the Checkbox's content
error                 | `string`           | `-`      | undefined   | Displays the Checkbox's error
autoFocus             | `bool`             | `-`      | false       | Makes the Checkbox's in focused
indeterminate         | `bool`             | `-`      | false       | Displays the indeterminate Checkbox 
multiple              | `bool`             | `-`      | false       | Displays the multiple Checkbox 
checked               | `bool`             | `-`      | false       | Displays the checked Checkbox 
disabled              | `bool`             | `-`      | false       | Displays the disabled Checkbox 
readOnly              | `bool`             | `-`      | false       | Displays the readOnly Checkbox 
hidden                | `bool`             | `-`      | false       | Makes the Checkbox's hidden
rendered              | `bool`             | `-`      | true        | Render Button to the DOM
className             | `string`           | `-`      | undefined   | Allows customizing styling of the component
labelClassName        | `string`           | `-`      | undefined   | Allows customizing styling of the Checkbox's label
innerRef              | `func`             | `-`      | undefined   | Get a reference to the DOM element
onChange              | `func`             | `-`      | undefined   | Runs when the value of the field changes


