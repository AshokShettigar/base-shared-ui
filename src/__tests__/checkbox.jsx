import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { CheckBox } from 'platform-common-ui/checkbox';

describe('checkbox', function() {
    const onChange = () => true;

    it('should render the checkbox', function() {
        const tree = renderer.create(
            <CheckBox id="ck1" onChange={onChange} value="ck1_value">Check me!</CheckBox>
        );
        expect(tree).toMatchSnapshot();
    });

    it('should render the checkbox with classname', function() {
        const tree = renderer.create(
            <CheckBox
                className="my-class"
                id="ck1"
                onChange={onChange}
                value="ck1_value"
            >
                Check me!
            </CheckBox>
        );
        expect(tree).toMatchSnapshot();
    });

    it('should render the checkbox with disabled', function() {
        const tree = renderer.create(
            <CheckBox id="ck1" isDisabled onChange={onChange} value="ck1_value">Check me!</CheckBox>
        );
        expect(tree).toMatchSnapshot();
    });

    it('should render the checkbox with no label', function() {
        const tree = renderer.create(
            <CheckBox id="ck1" hideLabel onChange={onChange} value="ck1_value">Check me!</CheckBox>
        );
        expect(tree).toMatchSnapshot();
    });

    it('should render the checkbox with checked', function() {
        const tree = renderer.create(
            <CheckBox id="ck1" isChecked onChange={onChange} value="ck1_value">Check me!</CheckBox>
        );
        expect(tree).toMatchSnapshot();
    });

    it('should trigger onChange handler', function(done) {
        const onChangeHandler = value => {
            expect(value).toBe('ck1_value');
            done();
        };

        const event = {
            target: { value: 'ck1_value' }
        };
        const wrapper = shallow(
            <CheckBox id="ck1" onChange={onChangeHandler} value="ck1_value">Check me!</CheckBox>
        );
        wrapper.find('input').simulate('change', event);
    });

    it('should throw console error when there is no children', function() {
        // eslint-disable-next-line no-console
        console.error = jest.fn();
        renderer.create(<CheckBox id="ck1" onChange={onChange} value="ck1_value" />);
        // eslint-disable-next-line no-console
        expect(console.error).toHaveBeenCalled();
    });
});
