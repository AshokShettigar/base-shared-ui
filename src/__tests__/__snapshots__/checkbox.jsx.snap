// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkbox should render the checkbox 1`] = `
<label
  className="cn-field__opt"
  data-component="CheckBox"
  htmlFor="ck1"
>
  <input
    checked={false}
    className="cn-field__checkradio"
    disabled={false}
    id="ck1"
    onChange={[Function]}
    type="checkbox"
    value="ck1_value"
  />
  <span
    className="cn-field__text"
  >
    Check me!
  </span>
</label>
`;

exports[`checkbox should render the checkbox with checked 1`] = `
<label
  className="cn-field__opt cn-field__opt--is-checked"
  data-component="CheckBox"
  htmlFor="ck1"
>
  <input
    checked={true}
    className="cn-field__checkradio"
    disabled={false}
    id="ck1"
    onChange={[Function]}
    type="checkbox"
    value="ck1_value"
  />
  <span
    className="cn-field__text"
  >
    Check me!
  </span>
</label>
`;

exports[`checkbox should render the checkbox with classname 1`] = `
<label
  className="cn-field__opt my-class"
  data-component="CheckBox"
  htmlFor="ck1"
>
  <input
    checked={false}
    className="cn-field__checkradio"
    disabled={false}
    id="ck1"
    onChange={[Function]}
    type="checkbox"
    value="ck1_value"
  />
  <span
    className="cn-field__text"
  >
    Check me!
  </span>
</label>
`;

exports[`checkbox should render the checkbox with disabled 1`] = `
<label
  className="cn-field__opt"
  data-component="CheckBox"
  htmlFor="ck1"
>
  <input
    checked={false}
    className="cn-field__checkradio"
    disabled={true}
    id="ck1"
    onChange={[Function]}
    type="checkbox"
    value="ck1_value"
  />
  <span
    className="cn-field__text"
  >
    Check me!
  </span>
</label>
`;

exports[`checkbox should render the checkbox with no label 1`] = `
<label
  className="cn-field__opt"
  data-component="CheckBox"
  htmlFor="ck1"
>
  <input
    checked={false}
    className="cn-field__checkradio"
    disabled={false}
    id="ck1"
    onChange={[Function]}
    type="checkbox"
    value="ck1_value"
  />
  <span
    className="cn-field__text a11y"
  >
    Check me!
  </span>
</label>
`;
